code/modules/xenoarchaeology/xenoartifact_console.dm

```diff
@@ -0,0 +1,256 @@
+	desc = "A science console used to source sellers, and buyers, for various blacklisted research objects."
+	icon_screen = "xenoartifact_console"
+	icon_keyboard = "rd_key"
+	circuit = /obj/item/circuitboard/computer/xenoartifact_console
+	
+	var/list/sellers = list() //These lengths need to be set to define an upper limit of sellers and buyers. Generally easier. 
+	var/list/buyers = list()
+	var/list/tab_index = list("Listings", "Export", "Linking") //All tabs
+	var/current_tab = "Listings"
+	var/current_tab_info = "Here you can find listings for various research samples, usually fresh from the field. These samples aren't distrubuted by the Nanotrasen affiliated cargo system, so instead listing data is sourced from stray bluespace-threads."
+	var/obj/machinery/xenoartifact_inbox/linked_inbox //used for 'shipping'
+	var/list/linked_machines = list() //List of linked machines for UI purposes
+	var/datum/techweb/linked_techweb //Which science server recieves points
+	var/list/sold_artifacts = list() //Actually just a general list of items you've sold, name is a legacy thing
+	var/datum/bank_account/budget //Which department's budget recieves profit
+
+/obj/machinery/computer/xenoartifact_console/Initialize()
+	. = ..()
+	linked_techweb = SSresearch.science_tech
+	budget = SSeconomy.get_dep_account(ACCOUNT_SCI)
+	sync_devices()
+	for(var/I in 1 to XENOA_MAX_VENDORS) //Add initial buyers and sellers
+		var/datum/xenoartifact_seller/S = new
+		sellers += S
+		S.generate()
+
+		var/datum/xenoartifact_seller/buyer/B = new
+		buyers += B
+		B.generate()
+
+/obj/machinery/computer/xenoartifact_console/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "XenoartifactConsole")
+		ui.open()
+
+/obj/machinery/computer/xenoartifact_console/ui_data(mob/user)
+	var/list/data = list()
+	data["points"] = budget ? budget.account_balance : 0
+	data["seller"] = list()
+	for(var/datum/xenoartifact_seller/S as() in sellers) //Pass seller data
+		if(S)
+			data["seller"] += list(list(
+				"name" = S.name,
+				"dialogue" = S.dialogue,
+				"price" = S.price,
+				"id" = S.unique_id,
+		))
+	data["buyer"] = list()
+	for(var/datum/xenoartifact_seller/buyer/B as() in buyers) //Buyer data
+		if(B)
+			data["buyer"] += list(list(
+				"name" = B.name,
+				"dialogue" = B.dialogue,
+				"price" = B.price,
+				"id" = B.unique_id,
+		))
+	data["tab_index"] = tab_index
+	data["current_tab"] = current_tab
+	data["tab_info"] = current_tab_info
+	data["linked_machines"] = linked_machines
+	data["sold_artifacts"] = sold_artifacts
+	return data
+
+/obj/machinery/computer/xenoartifact_console/ui_act(action, params) //I should probably use a switch statement for this but, the for statements look painful
+	. = TRUE
+	if(..())
+		return
+
+	if(action == "link_nearby")
+		sync_devices()
+		return
+	else if(action == "sell")
+		sell()
+		return
+
+	for(var/t in tab_index) //Handle swapping tabs.
+		if(action == "set_tab_[t]")
```

Comment:
this is cursed, you can just pass the arg with t instead of passing the action from tgui I believe